src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -21,7 +21,11 @@ namespace Microsoft.CodeAnalysis.LanguageServer
     internal static class Extensions
     {
         public static Uri GetURI(this TextDocument document)
-            => ProtocolConversions.GetUriFromFilePath(document.FilePath);
+        {
+            return document is SourceGeneratedDocument
+                ? ProtocolConversions.GetUriFromPartialFilePath(document.FilePath)
```

Comment:
Could this be done similar to metadata for goto def?  Where when we pass back the source generated URI it points to a real file on disk (in temp) that the client can open?

Otherwise I agree that the long term solution is to have a way for the server to provide content for the URI.  Am curious though - @NTaylorMullen do you happen to remember how this is done in VSCode?  IIRC this is handled outside of LSP through TextDocumentContentProvider, though I'm not sure why it couldn't be through an LSP method

Reply:
> @NTaylorMullen do you happen to remember how this is done in VSCode?

Yup! It's done in two ways in VSCode:

1. Server creates a temp file on disk with content and then navigates to that. Downside is that the file is mutable.
2. There's a client side component in VSCode that understands how to "provide content" for a given URI scheme. Meaning it registers for scheme "csharp-external://" and any URI that tries to navigate to anything under that scheme will first ask the provider for the content.

This all leads to an eventual of "how should this be done in LSP"? Answer is via a virtual file system expansion. [I started speccing](https://github.com/NTaylorMullen/LSPVirtualDocuments/blob/master/Documents/FileSystemSpec.md) out what that'd look like for embedded language purposes but it also is meant to handle this scenario too. Basically removes the client side portion of option 2 above and puts that responsibility into the server.

Reply:
> @dibarbet: Could this be done similar to metadata for goto def? Where when we pass back the source generated URI it points to a real file on disk (in temp) that the client can open?

Yeah, this could (and should) all work the same way.

> @NTaylorMullen: Answer is via a virtual file system expansion.

That would be an interesting approach. For the generator cases I can imagine that might fit well since we could make virtual folders/directories for different outputs. For other cases (like metadata as source) that might be a bit overkill. Would love to chat more about your thoughts, but that's now well out of scope of this PR.

Reply:
> Would love to chat more about your thoughts, but that's now well out of scope of this PR.

+100